core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -69,14 +83,15 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
     } catch {
       case e: FatalExitError =>
         isRunning.set(false)
-        shutdownLatch.countDown()
+        shutdownInitiated.countDown()
+        shutdownComplete.countDown()
         info("Stopped")
         Exit.exit(e.statusCode())
       case e: Throwable =>
         if (isRunning.get())
           error("Error due to", e)
     }
-    shutdownLatch.countDown()
+    shutdownComplete.countDown()
```

Comment:
I'm not sure it's strictly necessary, but it looks a bit clearer if this goes in a `finally` block.